stl/inc/xstring

```diff
@@ -2432,6 +2440,106 @@ public:
+            _Traits::copy(_Ptr + _My_data._Mysize, _Right_data._Myptr(), _Right_size);
+            _My_data._Mysize = _New_size;
+            _Traits::assign(_Ptr[_New_size], _Elem());
+            return;
+        }
+
+        if (_Allocators_equal(_Getal(), _Right._Getal()) && _Left_size <= _Right_capacity - _Right_size) {
+            // take _Right's buffer, max_size() is OK because the result fits in _Right's capacity
+            // At this point, we have tested:
+            // !(_Right_size <= _Left_capacity - _Left_size && _Left_capacity >= _Right_capacity)
+            //     && _Left_size <= _Right_capacity - _Right_size
+            // therefore: (demorgan)
+            // (_Right_size > _Left_capacity - _Left_size || _Left_capacity < _Right_capacity)
+            //     && _Left_size <= _Right_capacity - _Right_size
+            // therefore: (distribute)
+            // (_Right_size > _Left_capacity - _Left_size && _Left_size <= _Right_capacity - _Right_size)
+            // || (_Left_capacity < _Right_capacity && _Left_size <= _Right_capacity - _Right_size)
+            // therefore: either the result fits in _Right's capacity but not in _Left's capacity, so _Right must have
+            //     had more capacity, OR we explicitly tested that _Right had more capacity
+            // therefore: _Right must be in large mode (since it's capacity is larger than the minimum), so save that
+            //     branch
+            _STL_INTERNAL_CHECK(_Right_data._Large_string_engaged());
+            _My_data._Alloc_proxy(_GET_PROXY_ALLOCATOR(_Alty, _Getal())); // throws, hereafter nothrow in this block
+            _Take_contents(_Right, bool_constant<_Can_memcpy_val>{});
+            const auto _Ptr = _Unfancy(_My_data._Bx._Ptr);
+            _Traits::move(_Ptr + _Left_size, _Ptr, _My_data._Mysize);
+            _Traits::copy(_Ptr, _Left_data._Myptr(), _Left_size);
+            _My_data._Mysize = _New_size;
+            _Traits::assign(_Ptr[_New_size], _Elem());
+            return;
+        }
+
+        // can't use either buffer, reallocate
+        const auto _Max = max_size();
+        if (_Max - _Left_size < _Right_size) { // check if max_size() is OK
+            _Xlen_string();
+        }
+
+        const auto _New_capacity = _Calculate_growth(_New_size, _BUF_SIZE - 1, _Max);
+        auto&& _Alproxy          = _GET_PROXY_ALLOCATOR(_Alty, _Getal());
+        _Container_proxy_ptr<_Alty> _Proxy(_Alproxy, _My_data); // throws
+        const pointer _Fancyptr = _Getal().allocate(_New_capacity + 1); // throws
+        // nothrow hereafter
+        _Construct_in_place(_My_data._Bx._Ptr, _Fancyptr);
+        _My_data._Mysize = _New_size;
+        _My_data._Myres  = _New_capacity;
+        const auto _Ptr  = _Unfancy(_Fancyptr);
+        _Traits::copy(_Ptr, _Left_data._Myptr(), _Left_size);
+        _Traits::copy(_Ptr + static_cast<ptrdiff_t>(_Left_size), _Right_data._Myptr(), _Right_size);

```

Comment:
One more null terminator that could simply be copied along with the rest of the data.